### PR TITLE
Adding SimSystemModel class as higher level class containing HamiltonianModel

### PR DESCRIPTION
**Pull request by DanPuzzuoli**
### Update

This is no longer a [WIP] - and is ready for review.

### Summary

This is a [WIP], but the purpose of the new class `SimSystemModel` is to:
- Along the lines of the recent `HamiltonianModel` addition, detangle the options specifications for a pulse simulation. It contains a `HamiltonianModel` instance, along with any other remaining model information not naturally contained in `HamiltonianModel`, such as the sampling rate `dt` and `u_channel_lo`.
- The place that users can manipulate/probe the properties of the model; along with `HamiltonianModel`, this is where model-based functions like frequency calculations can live that the user can use.

To instantiate from a backend:

```
backend_real = FakeOpenPulse2Q()
system_model = SimSystemModel.from_backend(backend_real)
```
To simulate using this model, and an already assembled pulse `qobj`:
```
sim_result = backend_sim.run(qobj, system_model, backend_options=backend_options).result()
```
Now, the argument `backend_options` will contain only information specific to the simulator backend, like ODE options (as opposed to a combination of model and solver parameters as before).

### Details and comments

I still need to make the following changes, but comments are welcome.

- Change tests to use this object
- Change example/calibration notebooks to use this object
- Write docs for new functions,

I'm not crazy about the name `SimSystemModel`, any suggestions @chriseclectic @atilag ? `SystemModel` sounds better, but I thought it might be too general.

### CODE REVIEW

**Comment by DanPuzzuoli** [comment]
Update: Made some changes, but in terms of the to do list, migrated the `digest` test functions into a new file for testing behaviour of model classes.

There is a slight issue that will cause things to break at some point: `terra` now reports defaults in `s/Hz`, but some units in `configuration` for a backend seem to still be in `GHz`, including reporting of the Hamiltonian. This inconsistency makes things a bit awkward still; e.g. currently for things to work it's necessary to specify `test_model.dt = 1` after constructing from a backend, as the Hamiltonian parameters in `configuration` has units of GHz, but `configuration['dt']` is in seconds.

**Comment by DanPuzzuoli** [comment]
This PR functionally is now where I want it to be; it will be ready for the review process once I make sure the doc strings are fine and do linting/styling. 

Current summary of changes:

- Added `PulseSystemModel` object to store model information for calling the simulator with a `pulse` qobj
   - Stores a `HamiltonianModel`, which currently needs to be constructed from the hamiltonian string format, and contains functionality for computing dressed energies and what not
   - Has functions similar to the `PulseBackendConfiguration` for the purposes of 'calling' channels, e.g. `.drive(0)` to get `DriveChannel(0)`, so that everything can be done from the `PulseSystemModel` (for when there is no explicit backend that the model is based on)
   - Can be instantiated from a backend through the `PulseSystemModel.from_backend` constructor, but need not be (see the `TestPulseSimulator._system_model_1Q` method for a minimal specification of a `PulseSystemModel`)
- pulse simulator backend
   - minor changes to take `PulseSystemModel` as an argument
- `digest.py`
   - This file has been changed to accept the new objects, though functionally hasn't really changed. I've tried to re-organize the digest function to be a bit clearer where the parameters being passed to the simulator are being drawn from, though this could still be a lot cleaner.
- `hamiltonian_model.py`
  - put some parsing warnings in here
  - made a default constructor (one which just sets parameters), and one that accepts a hamiltonian string specification
  - put automatic determination of `qubit_list` from the qubit dimension dictionary.
  - renamed `_calculate_drift_hamiltonian` to `_compute_drift_data` as the function populates various fields based on the drift hamiltonian
  - added `set_vars` method, to allow changing of hamiltonian parameters
- deleted `pulse_sim_utils.py` as this functionality is now in the model objects
- added `test_system_models` as a place to put testing for the system model classes, and commented out the tests in `test_pulse_digest` (as they are now in `test_system_models`). Not sure if I should just delete `test_pulse_digest.py`, or just leave this file here to fill in later.

**Comment by DanPuzzuoli** [comment]
The latest updates are described by the previous comment; it has been edited as it is mostly the same. The only edit is the removal of the methods from `PulseSystemModel` that behave like a pulse backend configuration; when constructing schedules the current paradigm is for hte user to just directly call `pulse.DriveChannel(0)` to reference the drive channel on qubit 0.

File: qiskit/providers/aer/openpulse/pulse_system_model.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                 dt=None):
+        """Basic constructor.
+        """
+
+        # default type values
+        self.qubit_freq_est = qubit_freq_est
+        self.meas_freq_est = meas_freq_est
+
+        # necessary values
+        self.hamiltonian = hamiltonian
```
Comment (chriseclectic) [inline qiskit/providers/aer/openpulse/pulse_system_model.py]: If `hamiltonian` must be `None` or a `HamiltonianModel` object we should add a validation check:

```python
if hamiltonian is not None and not isinstance(hamiltonian, HamiltonianModel):
    raise AerError("hamiltonian must be a HamiltonianModel object")
```

File: qiskit/providers/aer/openpulse/pulse_system_model.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                raise ValueError("No qubit_lo_freq to use.")
+
+            qubit_lo_freq = self.qubit_freq_est
+
+        if self.u_channel_lo is None:
+            raise ValueError("{} has no u_channel_lo.".format(self.__class__.__name__))
+
+        # Setup freqs for the channels
+        freqs = OrderedDict()
+        for key in self.channels.keys():
```
Comment (chriseclectic) [inline qiskit/providers/aer/openpulse/pulse_system_model.py]: No need for keys()
```suggestion
        for key in self.channels:
```

File: qiskit/providers/aer/openpulse/pulse_system_model.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                 meas_freq_est=None,
+                 u_channel_lo=None,
+                 qubit_list=None,
+                 channels=None,
+                 dt=None):
+        """Basic constructor.
+        """
+
+        # default type values
+        self.qubit_freq_est = qubit_freq_est
```
Comment (chriseclectic) [inline qiskit/providers/aer/openpulse/pulse_system_model.py]: Not sure we should have public attributes for these. Maybe either store them in a `PulseDefaults` object that is returned by `.defaults` or have setter/getter methods for them

File: qiskit/providers/aer/backends/pulse_simulator.py [hunk 4]
```diff
@@ -57,18 +57,20 @@ def __init__(self, configuration=None, provider=None):
                          provider=provider)
 
     def run(self, qobj,
+            system_model,
             backend_options=None,
             noise_model=None,
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/backends/pulse_simulator.py]: Marked for future PRs.

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 5]
```diff
@@ -23,22 +24,37 @@
 
 class HamiltonianModel():
     """Hamiltonian model for pulse simulator."""
-    def __init__(self, hamiltonian, qubits=None):
+
+    def __init__(self,
+                 system=None,
+                 variables=None,
+                 dim_qub=None,
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Done

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 6]
```diff
@@ -48,100 +64,94 @@ def __init__(self, hamiltonian, qubits=None):
...
-                             "Hamiltonian if qubits list is not specified")
+        # populate self._channels
+        self._calculate_hamiltonian_channels()
+
+        # populate self._h_diag, self._evals, self._estates
+        self._compute_drift_data()
+
+    @classmethod
+    def from_string_spec(cls, hamiltonian, qubit_list=None):
+        """Initialize from a Hamiltonian string specification.
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Adding this to the list of things for future PRs.

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 7]
```diff
@@ -48,100 +64,94 @@ def __init__(self, hamiltonian, qubits=None):
...
 
-        self._vars = OrderedDict(hamiltonian['vars'])
+        # get variables
+        variables = OrderedDict(hamiltonian['vars'])
 
         # Get qubit subspace dimensions
         if 'qub' in hamiltonian:
-            self._dim_qub = {
+            if qubit_list is None:
+                qubit_list = [int(qubit) for qubit in hamiltonian['qub'].keys()]
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Done

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 8]
```diff
@@ -239,9 +261,11 @@ def _first_excited_state(qubit_idx, dim_qub):
     """
     vector = np.array([1.])
     # iterate through qubits, tensoring on the state
-    for qubit, dim in dim_qub.items():
-        new_vec = np.zeros(dim)
-        if int(qubit) == qubit_idx:
+    qubit_indices = [int(qubit) for qubit in dim_qub.keys()]
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Done

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 9]
```diff
@@ -48,100 +64,94 @@ def __init__(self, hamiltonian, qubits=None):
...
+        system.parse(qubit_list)
+        system = system.compiled
 
-        # Step #2: Determine Hamiltonian channels
-        self._calculate_hamiltonian_channels()
+        return cls(system, variables, dim_qub, dim_osc)
 
-        # Step 3: Calculate diagonal hamiltonian
-        self._calculate_drift_hamiltonian()
+    def set_variables(self, variables):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Removed

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 10]
```diff
@@ -23,22 +24,37 @@
...
+        # Qubit subspace dimensinos
+        self._dim_qub = dim_qub or {}
+        # Oscillator subspace dimensions
+        self._dim_osc = dim_osc or {}
+
+        # The rest are computed from the previous
+
         # These tell the order in which the channels are evaluated in
         # the RHS solver.
         self._channels = None
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Removed

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 11]
```diff
@@ -48,100 +64,94 @@ def __init__(self, hamiltonian, qubits=None):
...
+        min_eval = np.min(self._evals)
+        for q_idx in range(len(self._dim_qub)):
+            single_excite = _first_excited_state(q_idx, self._dim_qub)
+            dressed_eval = _eval_for_max_espace_overlap(
+                single_excite, self._evals, self._estates)
+            qubit_lo_freq[q_idx] = (dressed_eval - min_eval) / (2 * np.pi)
+
+        return qubit_lo_freq
 
     def _calculate_hamiltonian_channels(self):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Recapping discussion: This has to be in HamiltonianModel for now as it is part of the string parsing.

File: qiskit/providers/aer/openpulse/hamiltonian_model.py [hunk 12]
```diff
@@ -219,6 +229,18 @@ def _calculate_drift_hamiltonian(self):
...
+def _hamiltonian_parse_warnings(hamiltonian):
+    """Raises warnings for hamiltonian specification.
+
+    Parameters:
+        hamiltonian (dict): dictionary specification of hamiltonian
+    Returns:
+    Raises:
+    """
+    if 'osc' in hamiltonian:
+        warn('Oscillator-type systems are not supported.')
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/hamiltonian_model.py]: Done

File: qiskit/providers/aer/openpulse/pulse_system_model.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+        Returns:
+            PulseSystemModel: the PulseSystemModel constructed from the backend.
+
+        Raises:
+            ValueError: If channel or u_channel_lo are invalid.
+        """
+
+        # get relevant information from backend
+        defaults = backend.defaults().to_dict()
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/pulse_system_model.py]: Noted for later.

File: qiskit/providers/aer/openpulse/pulse_system_model.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+        # necessary values
+        self.hamiltonian = hamiltonian
+        if channels is None and hamiltonian is not None:
+            self.channels = hamiltonian._channels
+        self.u_channel_lo = u_channel_lo
+        self.qubit_list = qubit_list
+        self.dt = dt
+
+    @classmethod
+    def from_backend(cls, backend, qubit_list=None):
```
[the remark this answers is not in the document]
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/pulse_system_model.py]: Added the checking.

Have made a note for future PRs about `qubit_list` similar to previous comment

File: qiskit/providers/aer/openpulse/pulse_system_model.py (hunk 1 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/pulse_system_model.py]: done

File: qiskit/providers/aer/openpulse/pulse_system_model.py (hunk 3 above)
Comment (DanPuzzuoli) [reply qiskit/providers/aer/openpulse/pulse_system_model.py]: Marked to figure out later along with the `digest` stuff.

File: qiskit/providers/aer/backends/pulse_simulator.py [hunk 15]
```diff
@@ -9,7 +9,7 @@
 # Any modifications or derivative works of this code must retain this
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
-# pylint: disable=arguments-differ
+# pylint: disable=arguments-differ, missing-return-type-doc
```
Comment (chriseclectic) [inline qiskit/providers/aer/backends/pulse_simulator.py]: We shouldn't need to add this pylint skip, what functions were causing the `missing-return-type-doc` error?

**Pull request merged**